src/ILCompiler.Compiler/src/Compiler/TypeExtensions.cs

```diff
@@ -79,7 +79,7 @@ public static bool RequiresInstMethodTableArg(this MethodDesc method)
         /// </summary>
         public static bool AcquiresInstMethodTableFromThis(this MethodDesc method)
         {
-            return method.IsSharedByGenericInstantiations &&
+            return method.IsCanonicalMethod(CanonicalFormKind.Any) &&
```

Comment:
> In particular IsSharedByGenericInstantiations will say false on, say, `Foo<System.__Canon[]>`

That's weird ... I think it would be best to get rid of this method. (BTW: `MethodDesc::IsSharedByGenericInstantiations` would in your example would say true in CoreCLR.)